// nestjs-cli/bin/core/assets/builders/class-name.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const asset_enum_1 = require("../../../common/asset/enums/asset.enum");
class ClassNameBuilder {
    constructor() {
        this._name = null;
        this._asset = null;
    }
    addName(name) {
        this._name = name;
        return this;
    }
    addAsset(asset) {
        this._asset = asset;
        return this;
    }
    build() {
        return `${this.capitalize(this.mapName())}${this.mapAsset()}`;
    }
    mapAsset() {
        switch (this._asset) {
            case asset_enum_1.AssetEnum.MODULE:
                return 'Module';
            case asset_enum_1.AssetEnum.CONTROLLER:
                return 'Controller';
            case asset_enum_1.AssetEnum.COMPONENT:
                return 'Service';
        }
    }
    mapName() {
        if (this.isPath()) {
            return this.mapNameFromPath();
        }
        return this._name;
    }
    isPath() {
        return this._name.indexOf('/') !== -1;
    }
    mapNameFromPath() {
        const elements = this._name.split('/');
        return elements[elements.length - 1];
    }
    capitalize(name) {
        return `${name.charAt(0).toUpperCase()}${name.slice(1, name.length)}`;
    }
}
exports.ClassNameBuilder = ClassNameBuilder;
