// nestjs-cli/bin/core/assets/builders/file-name.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const asset_enum_1 = require("../../../common/asset/enums/asset.enum");
class FileNameBuilder {
    addName(name) {
        this._name = name;
        return this;
    }
    addAsset(asset) {
        this._asset = asset;
        return this;
    }
    addTest(isTest) {
        this._isTest = isTest;
        return this;
    }
    addExtension(extension) {
        this._extension = extension;
        return this;
    }
    build() {
        return `${this.mapName()}.${this.mapAsset()}${this.mapTest()}.${this._extension}`;
    }
    mapTest() {
        return this._isTest ? '.spec' : '';
    }
    mapAsset() {
        switch (this._asset) {
            case asset_enum_1.AssetEnum.MODULE:
                return 'module';
            case asset_enum_1.AssetEnum.CONTROLLER:
                return 'controller';
            case asset_enum_1.AssetEnum.COMPONENT:
                return 'service';
        }
    }
    mapName() {
        if (this.isPath()) {
            return this.mapNameFromPath();
        }
        return this._name;
    }
    isPath() {
        return this._name.indexOf('/') !== -1;
    }
    mapNameFromPath() {
        const elements = this._name.split('/');
        return elements[elements.length - 1];
    }
}
exports.FileNameBuilder = FileNameBuilder;
