// nestjs-cli/bin/core/assets/processors/component.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const asset_generator_1 = require("../generators/asset.generator");
const path = require("path");
const template_builder_1 = require("../builders/template.builder");
const asset_enum_1 = require("../../../common/asset/enums/asset.enum");
const file_name_builder_1 = require("../builders/file-name.builder");
const asset_builder_1 = require("../builders/asset.builder");
const class_name_builder_1 = require("../builders/class-name.builder");
const module_updater_1 = require("../module-updaters/module.updater");
class ComponentProcessor {
    constructor(_name, _extension) {
        this._name = _name;
        this._extension = _extension;
        this._generator = new asset_generator_1.AssetGenerator();
        this._updater = new module_updater_1.ModuleUpdaterImpl();
        this._assets = [];
    }
    process() {
        this.buildAssets();
        return this.generate()
            .then(() => this.updateModule());
    }
    buildAssets() {
        const className = new class_name_builder_1.ClassNameBuilder()
            .addName(this._name)
            .addAsset(asset_enum_1.AssetEnum.COMPONENT)
            .build();
        const filename = new file_name_builder_1.FileNameBuilder()
            .addAsset(asset_enum_1.AssetEnum.COMPONENT)
            .addName(this._name)
            .addExtension(this._extension)
            .addTest(false)
            .build();
        this._assets.push(this.buildClassAsset(className, filename));
        this._assets.push(this.buildTestAsset(className, filename));
    }
    buildClassAsset(className, filename) {
        return new asset_builder_1.AssetBuilder()
            .addClassName(className)
            .addFilename(path.join(process.cwd(), 'src/app/modules', this._name, filename))
            .addTemplate(new template_builder_1.TemplateBuilder()
            .addFilename(path.resolve(__dirname, `../../../assets/${this._extension}/component/component.${this._extension}.template`))
            .addReplacer({
            __CLASS_NAME__: className
        })
            .build())
            .build();
    }
    buildTestAsset(className, filename) {
        return new asset_builder_1.AssetBuilder()
            .addClassName(className)
            .addFilename(path.join(process.cwd(), 'src/app/modules', this._name, new file_name_builder_1.FileNameBuilder()
            .addAsset(asset_enum_1.AssetEnum.COMPONENT)
            .addName(this._name)
            .addExtension(this._extension)
            .addTest(true)
            .build()))
            .addTemplate(new template_builder_1.TemplateBuilder()
            .addFilename(path.resolve(__dirname, `../../../assets/${this._extension}/component/component.spec.${this._extension}.template`))
            .addReplacer({
            __CLASS_NAME__: className,
            __IMPORT__: filename
        })
            .build())
            .build();
    }
    generate() {
        return this._generator.generate(this._assets[0])
            .then(() => this._generator.generate(this._assets[1]));
    }
    updateModule() {
        return this._updater.update(this._assets[0].filename, this._assets[0].className, asset_enum_1.AssetEnum.COMPONENT);
    }
}
exports.ComponentProcessor = ComponentProcessor;
